src/include/btree_inline.h

```diff
@@ -2398,3 +2398,139 @@ __wt_btcur_skip_page(
     WT_REF_UNLOCK(ref, previous_state);
     return (0);
 }
+
+/*
+ * __wt_ref_index_slot --
+ *     Return the page's index and slot for a reference.
+ */
+static WT_INLINE void
+__wt_ref_index_slot(WT_SESSION_IMPL *session, WT_REF *ref, WT_PAGE_INDEX **pindexp, uint32_t *slotp)
+{
+    WT_PAGE_INDEX *pindex;
+    WT_REF **p, **start, **stop, **t;
+    uint64_t sleep_usecs, yield_count;
+    uint32_t entries, slot;
+
+    /*
+     * If we don't find our reference, the page split and our home pointer references the wrong
+     * page. When internal pages split, their WT_REF structure home values are updated; yield and
+     * wait for that to happen.
+     */
+    for (sleep_usecs = yield_count = 0;;) {
+        /*
+         * Copy the parent page's index value: the page can split at any time, but the index's value
+         * is always valid, even if it's not up-to-date.
+         */
+        WT_INTL_INDEX_GET(session, ref->home, pindex);
+        entries = pindex->entries;
+
+        /*
+         * Use the page's reference hint: it should be correct unless there was a split or delete in
+         * the parent before our slot. If the hint is wrong, it can be either too big or too small,
+         * but often only by a small amount. Search up and down the index starting from the hint.
+         *
+         * It's not an error for the reference hint to be wrong, it just means the first retrieval
+         * (which sets the hint for subsequent retrievals), is slower.
+         */
+        slot = ref->pindex_hint;
+        if (slot >= entries)
+            slot = entries - 1;
+        if (pindex->index[slot] == ref)
+            goto found;
+        for (start = &pindex->index[0], stop = &pindex->index[entries - 1],
+            p = t = &pindex->index[slot];
+             p > start || t < stop;) {
+            if (p > start && *--p == ref) {
+                slot = (uint32_t)(p - start);
+                goto found;
+            }
+            if (t < stop && *++t == ref) {
+                slot = (uint32_t)(t - start);
+                goto found;
+            }
+        }
+        /*
+         * We failed to get the page index and slot reference, yield before retrying, and if we've
+         * yielded enough times, start sleeping so we don't burn CPU to no purpose.
+         */
+        __wt_spin_backoff(&yield_count, &sleep_usecs);
+        WT_STAT_CONN_INCRV(session, page_index_slot_ref_blocked, sleep_usecs);
+    }
+
+found:
+    WT_ASSERT(session, pindex->index[slot] == ref);
+    *pindexp = pindex;
+    *slotp = slot;
+}
+
+/*
+ * __wt_ref_ascend --
+ *     Ascend the tree one level. If `pindexp` is not null, fill in both `pindexp` and `slotp`.
+ */
+static WT_INLINE void
+__wt_ref_ascend(WT_SESSION_IMPL *session, WT_REF **refp, WT_PAGE_INDEX **pindexp, uint32_t *slotp)
```

Comment:
Are the two functions pulled into here identical except for __wt?

Reply:
yes, it's moved over